man/man3/Logger.3
.TH "Logger" 3 "Version 0.1" "Max OS" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Logger \- A class that handles logging messages to the console and files\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <logger\&.h>\fP
.PP
Inherits \fBMaxOS::common::OutputStream\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBvoid\fP \fBadd_log_writer\fP (\fBOutputStream\fP *\fBlog_writer\fP)"
.br
.RI "Adds an output stream to the logger\&. "
.ti -1c
.RI "\fBvoid\fP \fBdisable_log_writer\fP (\fBOutputStream\fP *\fBlog_writer\fP)"
.br
.RI "Removes an output stream from the logger\&. "
.ti -1c
.RI "\fBvoid\fP \fBset_log_level\fP (\fBLogLevel\fP \fBlog_level\fP)"
.br
.RI "Sets the log level of the logger\&. "
.ti -1c
.RI "\fBvoid\fP \fBwrite_char\fP (\fBchar\fP c) \fBfinal\fP"
.br
.RI "Writes a character to the logger\&. "
.ti -1c
.RI "\fBvoid\fP \fBprintf\fP (\fBconst\fP \fBchar\fP *format,\&.\&.\&.)"
.br
.RI "Prints a formatted string to the logger\&. "
.ti -1c
.RI "\fBLogger\fP & \fBoperator<<\fP (\fBLogLevel\fP \fBlog_level\fP)"
.br
.RI "Sets the log level of the logger\&. "
.in -1c

Public Member Functions inherited from \fBMaxOS::common::OutputStream\fP
.in +1c
.ti -1c
.RI "\fBvirtual\fP \fBvoid\fP \fBline_feed\fP ()"
.br
.RI "Writes a newline to the output stream\&. "
.ti -1c
.RI "\fBvirtual\fP \fBvoid\fP \fBcarriageReturn\fP ()"
.br
.RI "Writes a carriage return to the output stream\&. "
.ti -1c
.RI "\fBvirtual\fP \fBvoid\fP \fBclear\fP ()"
.br
.RI "Clears the output stream\&. "
.ti -1c
.RI "\fBvoid\fP \fBwrite\fP (\fBstring\fP \fBstring_to_write\fP) \fBoverride\fP"
.br
.RI "Writes a string to the output stream\&. "
.ti -1c
.RI "\fBvoid\fP \fBwrite\fP (\fBconst\fP \fBchar\fP *\fBstring_to_write\fP)"
.br
.RI "Writes a string to the output stream\&. "
.ti -1c
.RI "\fBvirtual\fP \fBvoid\fP \fBwrite_int\fP (\fBint\fP \fBint_to_write\fP)"
.br
.RI "Writes an integer to the output stream\&. "
.ti -1c
.RI "\fBvirtual\fP \fBvoid\fP \fBwrite_hex\fP (\fBuint64_t\fP \fBhex_to_write\fP)"
.br
.RI "Writes a hex to the output stream\&. "
.ti -1c
.RI "\fBOutputStream\fP & \fBoperator<<\fP (\fBstring\fP \fBstring_to_write\fP) \fBoverride\fP"
.br
.RI "Writes a string to the output stream\&. "
.ti -1c
.RI "\fBOutputStream\fP & \fBoperator<<\fP (\fBconst\fP \fBchar\fP *\fBstring_to_write\fP)"
.br
.RI "Writes a string to the output stream\&. "
.ti -1c
.RI "\fBOutputStream\fP & \fBoperator<<\fP (\fBint\fP \fBint_to_write\fP)"
.br
.RI "Writes a integer to the output stream\&. "
.ti -1c
.RI "\fBOutputStream\fP & \fBoperator<<\fP (\fBuint64_t\fP \fBhex_to_write\fP)"
.br
.RI "Writes a hexadecimal to the output stream\&. "
.ti -1c
.RI "\fBOutputStream\fP & \fBoperator<<\fP (\fBchar\fP \fBchar_to_write\fP)"
.br
.RI "Writes a character to the output stream\&. "
.in -1c

Public Member Functions inherited from \fBMaxOS::common::GenericOutputStream< string >\fP
.in +1c
.ti -1c
.RI "\fBGenericOutputStream\fP ()"
.br
.RI "__________________________________________Templates__________________________________________________/// "
.ti -1c
.RI "\fB~GenericOutputStream\fP ()"
.br
.RI "Destructor of the \fBGenericOutputStream\fP class\&. "
.ti -1c
.RI "\fBvoid\fP \fBon_stream_read\fP (\fBstring\fP) \fBoverride\fP"
.br
.RI "Writes the date that was read from the input stream to the output stream\&. "
.ti -1c
.RI "\fBvoid\fP \fBon_end_of_stream\fP (\fBGenericInputStream\fP< \fBstring\fP > *) \fBoverride\fP"
.br
.RI "Close the stream and remove it from the list of streams when the end of the stream is reached\&. "
.ti -1c
.RI "\fBvirtual\fP \fBvoid\fP \fBwrite\fP (\fBstring\fP)"
.br
.RI "write an element to the stream\&. "
.ti -1c
.RI "\fBvirtual\fP \fBvoid\fP \fBclose\fP ()"
.br
.RI "Close the stream\&. "
.ti -1c
.RI "\fBvirtual\fP \fBGenericOutputStream\fP< \fBstring\fP > & \fBoperator<<\fP (\fBstring\fP)"
.br
.RI "Overload the << operator to write an element to the stream\&. "
.in -1c

Public Member Functions inherited from \fBMaxOS::common::InputStreamProcessor< Type, ProcessorType >\fP
.in +1c
.ti -1c
.RI "\fBInputStreamProcessor\fP ()"
.br
.RI "Creates a new \fBInputStreamProcessor\fP\&. "
.ti -1c
.RI "\fBInputStreamProcessor\fP (\fBInputStreamEventHandler\fP< \fBProcessorType\fP > *\fBgeneric_stream_event_handler\fP)"
.br
.RI "Creates a new \fBInputStreamProcessor\fP\&. "
.ti -1c
.RI "\fB~InputStreamProcessor\fP ()"
.br
.RI "Destroys the \fBInputStreamProcessor\fP\&. "
.ti -1c
.RI "\fBvoid\fP \fBon_end_of_stream\fP (\fBGenericInputStream\fP< \fBType\fP > *\fBstream\fP) \fBoverride\fP"
.br
.RI "Called when a stream has finished\&. Passes the event on to the handlers and then removes the stream from the array of streams\&. "
.in -1c

Public Member Functions inherited from \fBMaxOS::common::InputStreamEventHandler< Type >\fP
.in +1c
.ti -1c
.RI "\fBInputStreamEventHandler\fP ()"
.br
.RI "_______________________________________________TEMPLATES_________________________________________________________________/// "
.ti -1c
.RI "\fB~InputStreamEventHandler\fP ()"
.br
.RI "Destroys the \fBInputStreamProcessor\fP and disconnects it from all streams\&. "
.in -1c

Public Member Functions inherited from \fBMaxOS::common::GenericInputStream< ProcessorType >\fP
.in +1c
.ti -1c
.RI "\fBGenericInputStream\fP ()"
.br
.RI "Creates a new \fBGenericInputStream\fP\&. "
.ti -1c
.RI "\fBGenericInputStream\fP (\fBInputStreamEventHandler\fP< \fBProcessorType\fP > *)"
.br
.RI "Creates a new \fBGenericInputStream\fP and connects it to the handler\&. "
.ti -1c
.RI "\fB~GenericInputStream\fP ()"
.br
.RI "Destroys the \fBGenericInputStream\fP and disconnects all handlers\&. "
.ti -1c
.RI "\fBvoid\fP \fBconnect_input_stream_event_handler\fP (\fBInputStreamEventHandler\fP< \fBProcessorType\fP > *)"
.br
.RI "Adds a inputStreamEventHandler to the list of internetProtocolHandlers\&. "
.ti -1c
.RI "\fBvoid\fP \fBdisconnect_input_stream_event_handler\fP (\fBInputStreamEventHandler\fP< \fBProcessorType\fP > *)"
.br
.RI "Removes a handler from the list of handlers\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "\fBstatic\fP \fBvoid\fP \fBASSERT\fP (\fBbool\fP \fBcondition\fP, \fBconst\fP \fBchar\fP *\fBmessage\fP,\&.\&.\&.)"
.br
.RI "Puts the system into a panic state if the condition is false, printing the message first\&. "
.ti -1c
.RI "\fBstatic\fP \fBLogger\fP * \fBactive_logger\fP ()"
.br
.RI "Gets the active logger (to be used to modify the logger) "
.ti -1c
.RI "\fBstatic\fP \fBLogger\fP & \fBOut\fP ()"
.br
.RI "Gets the active logger\&. "
.ti -1c
.RI "\fBstatic\fP \fBLogger\fP \fBHEADER\fP ()"
.br
.RI "Gets active logger set to task level\&. "
.ti -1c
.RI "\fBstatic\fP \fBLogger\fP \fBINFO\fP ()"
.br
.RI "Gets active logger set to info level\&. "
.ti -1c
.RI "\fBstatic\fP \fBLogger\fP \fBDEBUG\fP ()"
.br
.RI "Gets active logger set to DEBUG level\&. "
.ti -1c
.RI "\fBstatic\fP \fBLogger\fP \fBWARNING\fP ()"
.br
.RI "Gets active logger set to WARNING level\&. "
.ti -1c
.RI "\fBstatic\fP \fBLogger\fP \fBERROR\fP ()"
.br
.RI "Gets active logger set to ERROR level\&. "
.in -1c
.SS "Additional Inherited Members"


Protected Attributes inherited from \fBMaxOS::common::InputStreamEventHandler< Type >\fP
.in +1c
.ti -1c
.RI "\fBcommon::Vector\fP< \fBGenericInputStream\fP< \fBType\fP > * > \fBm_generic_input_streams\fP"
.br
.RI "List of streams being observed by this handler\&. "
.in -1c

Protected Attributes inherited from \fBMaxOS::common::GenericInputStream< ProcessorType >\fP
.in +1c
.ti -1c
.RI "\fBcommon::Vector\fP< \fBInputStreamEventHandler\fP< \fBProcessorType\fP > * > \fBm_input_stream_event_handlers\fP"
.br
.RI "List of streams being observed by this handler\&. "
.in -1c
.SH "Detailed Description"
.PP 
A class that handles logging messages to the console and files\&. 
.PP
Definition at line \fB39\fP of file \fBlogger\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Logger::Logger ()"

.PP
Definition at line \fB22\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
23 : m_log_writers()
24 {
25 
26     s_active_logger = this;
27 
28     // The following line is generated automatically by the MaxOS build system\&.
29     s_progress_total = 22;
30 
31 }
.fi

.SS "Logger::~Logger ()"

.PP
Definition at line \fB34\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
34                 {
35 
36 
37 }
.fi

.SH "Member Function Documentation"
.PP 
.SS "\fBLogger\fP * Logger::active_logger ()\fC [static]\fP"

.PP
Gets the active logger (to be used to modify the logger) 
.PP
\fBReturns\fP
.RS 4
The active logger 
.RE
.PP

.PP
Definition at line \fB213\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
213                               {
214     return s_active_logger;
215 }
.fi

.PP
Referenced by \fBMaxOS::drivers::console::VESABootConsole::finish()\fP, \fBOut()\fP, and \fBMaxOS::drivers::console::VESABootConsole::VESABootConsole()\fP\&.
.SS "\fBvoid\fP Logger::add_log_writer (\fBOutputStream\fP * log_writer)"

.PP
Adds an output stream to the logger\&. 
.PP
\fBParameters\fP
.RS 4
\fIlog_writer\fP The output stream to add 
.RE
.PP

.PP
Definition at line \fB44\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
44                                                     {
45 
46     // If the list is not empty
47     if (m_log_writer_count >= MAX_LOG_WRITERS)
48         return;
49 
50     // Add the output stream to the list
51     m_log_writers[m_log_writer_count] = log_writer;
52     m_log_writers_enabled[m_log_writer_count] = true;
53     m_log_writer_count++;
54 
55     // Print the setup info
56     *this << LogLevel::INFO << "Logger setup: " << m_log_writer_count << "\\n";
57     *this << LogLevel::HEADER << "MaxOS v" << VERSION_STRING << " [ build " << BUILD_NUMBER << " ]\\n";
58 
59 }
.fi

.PP
References \fBMAX_LOG_WRITERS\fP\&.
.PP
Referenced by \fBMaxOS::drivers::SerialConsole::SerialConsole()\fP\&.
.SS "\fBvoid\fP Logger::ASSERT (\fBbool\fP condition, \fBconst\fP \fBchar\fP * message,  \&.\&.\&.)\fC [static]\fP"

.PP
Puts the system into a panic state if the condition is false, printing the message first\&. 
.PP
\fBParameters\fP
.RS 4
\fIcondition\fP The condition to check if is met 
.br
\fImessage\fP The message to print if the condition fails 
.RE
.PP

.PP
Definition at line \fB238\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
238                                                             {
239 
240     // If the condition is met then everything is ok
241     if (condition)
242         return;
243 
244     va_list parameters;
245     va_start(parameters, message);
246 
247     // Hang the system
248     system::CPU::PANIC(String::formatted(message, parameters)\&.c_str());
249 }
.fi

.PP
References \fBMaxOS::String::formatted()\fP, and \fBMaxOS::system::CPU::PANIC()\fP\&.
.SS "\fBLogger\fP Logger::DEBUG ()\fC [static]\fP"

.PP
Gets active logger set to DEBUG level\&. 
.PP
\fBReturns\fP
.RS 4
The debug logger 
.RE
.PP

.PP
Definition at line \fB178\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
178                      {
179 
180     s_active_logger\->set_log_level(LogLevel::DEBUG);
181     return Out();
182 
183 }
.fi

.PP
References \fBOut()\fP\&.
.PP
Referenced by \fBMaxOS::processes::GlobalScheduler::add_process()\fP, \fBMaxOS::processes::GlobalScheduler::add_thread()\fP, \fBMaxOS::hardwarecommunication::AdvancedConfigurationAndPowerInterface::AdvancedConfigurationAndPowerInterface()\fP, \fBMaxOS::hardwarecommunication::AdvancedProgrammableInterruptController::AdvancedProgrammableInterruptController()\fP, \fBMaxOS::drivers::clock::Clock::calibrate()\fP, \fBMaxOS::system::CPU::check_nx()\fP, \fBMaxOS::system::Core::Core()\fP, \fBcore_main()\fP, \fBMaxOS::system::Core::init_sse()\fP, \fBMaxOS::system::Core::init_tss()\fP, \fBMaxOS::drivers::disk::IntegratedDriveElectronicsController::initialise()\fP, \fBMaxOS::hardwarecommunication::IOAPIC::IOAPIC()\fP, \fBMaxOS::system::GlobalDescriptorTable::load()\fP, \fBMaxOS::processes::GlobalScheduler::load_multiboot_elfs()\fP, \fBMaxOS::filesystem::partition::MSDOSPartition::mount_partitions()\fP, \fBMaxOS::system::Multiboot::Multiboot()\fP, \fBMaxOS::memory::PhysicalMemoryManager::PhysicalMemoryManager()\fP, \fBMaxOS::memory::PhysicalMemoryManager::reserve()\fP, \fBMaxOS::hardwarecommunication::PeripheralComponentInterconnectController::select_drivers()\fP, \fBMaxOS::drivers::video::VideoElectronicsStandardsAssociation::VideoElectronicsStandardsAssociation()\fP, \fBMaxOS::system::Core::wake_up()\fP, and \fBMaxOS::processes::Process::~Process()\fP\&.
.SS "\fBvoid\fP Logger::disable_log_writer (\fBOutputStream\fP * log_writer)"

.PP
Removes an output stream from the logger\&. 
.PP
\fBParameters\fP
.RS 4
\fIlog_writer\fP The output stream to remove 
.RE
.PP

.PP
Definition at line \fB66\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
66                                                         {
67 
68     // If the list is empty
69     if (m_log_writer_count == 0)
70         return;
71 
72     // Find the output stream in the list
73     for (int i = 0; i < m_log_writer_count; i++)
74         if (m_log_writers[i] == log_writer)
75             m_log_writers_enabled[i] = false;
76 }
.fi

.SS "\fBLogger\fP Logger::ERROR ()\fC [static]\fP"

.PP
Gets active logger set to ERROR level\&. 
.PP
\fBReturns\fP
.RS 4
The error logger 
.RE
.PP

.PP
Definition at line \fB201\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
201                      {
202 
203     s_active_logger\->set_log_level(LogLevel::ERROR);
204     return Out();
205 
206 }
.fi

.PP
References \fBOut()\fP\&.
.PP
Referenced by \fBMaxOS::filesystem::format::Fat32Volume::Fat32Volume()\fP, \fBMaxOS::system::SyscallManager::handle_interrupt()\fP, \fBMaxOS::system::CPU::PANIC()\fP, \fBMaxOS::system::CPU::prepare_for_panic()\fP, \fBMaxOS::system::CPU::print_registers()\fP, and \fBMaxOS::system::CPU::stack_trace()\fP\&.
.SS "\fBLogger\fP Logger::HEADER ()\fC [static]\fP"

.PP
Gets active logger set to task level\&. 
.PP
\fBReturns\fP
.RS 4
The task logger 
.RE
.PP

.PP
Definition at line \fB155\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
155                       {
156 
157     s_active_logger\->set_log_level(LogLevel::HEADER);
158     return Out();
159 
160 }
.fi

.PP
References \fBOut()\fP\&.
.PP
Referenced by \fBMaxOS::drivers::console::VESABootConsole::finish()\fP, and \fBkernel_main()\fP\&.
.SS "\fBLogger\fP Logger::INFO ()\fC [static]\fP"

.PP
Gets active logger set to info level\&. 
.PP
\fBReturns\fP
.RS 4
The info logger 
.RE
.PP

.PP
Definition at line \fB167\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
167                     {
168 
169     s_active_logger\->set_log_level(LogLevel::INFO);
170     return Out();
171 }
.fi

.PP
References \fBOut()\fP\&.
.PP
Referenced by \fBMaxOS::hardwarecommunication::InterruptManager::activate()\fP, \fBMaxOS::drivers::DriverManager::activate_drivers()\fP, \fBMaxOS::hardwarecommunication::AdvancedConfigurationAndPowerInterface::AdvancedConfigurationAndPowerInterface()\fP, \fBMaxOS::hardwarecommunication::AdvancedProgrammableInterruptController::AdvancedProgrammableInterruptController()\fP, \fBMaxOS::drivers::clock::Clock::calibrate()\fP, \fBMaxOS::drivers::clock::Clock::Clock()\fP, \fBMaxOS::system::CPU::CPU()\fP, \fBMaxOS::drivers::DriverManager::find_drivers()\fP, \fBMaxOS::processes::GlobalScheduler::GlobalScheduler()\fP, \fBMaxOS::system::CPU::init_cores()\fP, \fBMaxOS::drivers::DriverManager::initialise_drivers()\fP, \fBkernel_main()\fP, \fBMaxOS::memory::PhysicalMemoryManager::PhysicalMemoryManager()\fP, \fBMaxOS::drivers::DriverManager::reset_devices()\fP, \fBMaxOS::system::SyscallManager::syscall_klog()\fP, \fBMaxOS::drivers::console::VESABootConsole::VESABootConsole()\fP, and \fBMaxOS::drivers::video::VideoElectronicsStandardsAssociation::VideoElectronicsStandardsAssociation()\fP\&.
.SS "\fBLogger\fP & Logger::operator<< (\fBLogLevel\fP log_level)"

.PP
Sets the log level of the logger\&. 
.PP
\fBParameters\fP
.RS 4
\fIlog_level\fP The log level to set 
.RE
.PP
\fBReturns\fP
.RS 4
This logger 
.RE
.PP

.PP
Definition at line \fB257\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
257                                              {
258 
259     set_log_level(log_level);
260     return *this;
261 
262 }
.fi

.PP
References \fBset_log_level()\fP\&.
.SS "\fBLogger\fP & Logger::Out ()\fC [static]\fP"

.PP
Gets the active logger\&. 
.PP
\fBReturns\fP
.RS 4
The active logger 
.RE
.PP

.PP
Definition at line \fB145\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
145                     {
146 
147     return *active_logger();
148 }
.fi

.PP
References \fBactive_logger()\fP\&.
.PP
Referenced by \fBDEBUG()\fP, \fBERROR()\fP, \fBHEADER()\fP, \fBINFO()\fP, \fBMaxOS::hardwarecommunication::PeripheralComponentInterconnectController::list_known_device()\fP, \fBMaxOS::filesystem::partition::MSDOSPartition::mount_partitions()\fP, \fBMaxOS::processes::GlobalScheduler::print_running_header()\fP, \fBMaxOS::hardwarecommunication::PeripheralComponentInterconnectController::select_drivers()\fP, \fBMaxOS::system::SyscallManager::syscall_klog()\fP, and \fBWARNING()\fP\&.
.SS "\fBvoid\fP Logger::printf (\fBconst\fP \fBchar\fP * format,  \&.\&.\&.)"

.PP
Prints a formatted string to the logger\&. 
.PP
\fBParameters\fP
.RS 4
\fIformat\fP The format string 
.br
\fI\&.\&.\&.\fP The arguments to format 
.RE
.PP

.PP
Definition at line \fB223\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
223                                            {
224 
225     // Create a pointer to the data
226     va_list parameters;
227     va_start(parameters, format);
228 
229     write(String::formatted(format, parameters));
230 }
.fi

.PP
References \fBMaxOS::String::formatted()\fP, and \fBwrite\fP\&.
.SS "\fBvoid\fP Logger::set_log_level (\fBLogLevel\fP log_level)"

.PP
Sets the log level of the logger\&. 
.PP
\fBParameters\fP
.RS 4
\fIlog_level\fP The log level to set 
.RE
.PP

.PP
Definition at line \fB83\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
83                                              {
84 
85     // Set the log level
86     m_log_level = log_level;
87 
88     // Update the progress bar
89     if (log_level == LogLevel::INFO) {
90         VESABootConsole::update_progress_bar((m_progress_current * 100) / s_progress_total);
91         m_progress_current++;
92     }
93 
94     // Print the header
95     switch (log_level) {
96 
97         case LogLevel::HEADER:
98             *this << ANSI_COLOURS[ANSIColour::FG_Blue] << "[  BOOT    ] ";
99             break;
100 
101         case LogLevel::INFO:
102             *this << ANSI_COLOURS[ANSIColour::FG_Cyan] << "[  INFO    ]" << ANSI_COLOURS[ANSIColour::FG_White] << " ";
103             break;
104 
105         case LogLevel::DEBUG:
106             *this << ANSI_COLOURS[ANSIColour::FG_Yellow] << "[  DEBUG   ]" << ANSI_COLOURS[ANSIColour::Reset] << " ";
107             break;
108 
109         case LogLevel::WARNING:
110             *this << ANSI_COLOURS[ANSIColour::BG_Yellow] << ANSI_COLOURS[FG_White] << "[ WARNING  ]" << ANSI_COLOURS[ANSIColour::Reset] << " ";
111             break;
112 
113         case LogLevel::ERROR:
114             *this << ANSI_COLOURS[ANSIColour::BG_Red] << "[  ERROR   ]" << ANSI_COLOURS[ANSIColour::Reset] << " ";
115             break;
116     }
117 
118     GlobalScheduler::print_running_header();
119 }
.fi

.PP
References \fBMaxOS::common::ANSI_COLOURS\fP, \fBMaxOS::processes::GlobalScheduler::print_running_header()\fP, and \fBMaxOS::drivers::console::VESABootConsole::update_progress_bar()\fP\&.
.PP
Referenced by \fBoperator<<()\fP\&.
.SS "\fBLogger\fP Logger::WARNING ()\fC [static]\fP"

.PP
Gets active logger set to WARNING level\&. 
.PP
\fBReturns\fP
.RS 4
The warning logger 
.RE
.PP

.PP
Definition at line \fB190\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
190                        {
191 
192     s_active_logger\->set_log_level(LogLevel::WARNING);
193     return Out();
194 }
.fi

.PP
References \fBOut()\fP\&.
.PP
Referenced by \fBMaxOS::drivers::ethernet::AMD_AM79C973::handle_interrupt()\fP, \fBMaxOS::hardwarecommunication::InterruptManager::handle_interrupt_request()\fP, \fBMaxOS::drivers::disk::AdvancedTechnologyAttachment::identify()\fP, \fBMaxOS::filesystem::VirtualFileSystem::mount_filesystem()\fP, \fBMaxOS::filesystem::VirtualFileSystem::mount_filesystem()\fP, \fBMaxOS::filesystem::partition::MSDOSPartition::mount_partitions()\fP, \fBMaxOS::filesystem::VirtualFileSystem::unmount_filesystem()\fP, and \fBMaxOS::system::Core::wake_up()\fP\&.
.SS "\fBvoid\fP Logger::write_char (\fBchar\fP c)\fC [final]\fP, \fC [virtual]\fP"

.PP
Writes a character to the logger\&. 
.PP
\fBParameters\fP
.RS 4
\fIc\fP The character to write 
.RE
.PP

.PP
Reimplemented from \fBMaxOS::common::OutputStream\fP\&.
.PP
Definition at line \fB127\fP of file \fBlogger\&.cpp\fP\&..PP
.nf
127                               {
128 
129     // Ensure logging at this level is enabled
130     if (m_log_level > MAX_LOG_LEVEL)
131         return;
132 
133     // Write the character to all output streams
134     for (int i = 0; i < m_log_writer_count; i++)
135         if (m_log_writers_enabled[i])
136             m_log_writers[i]\->write_char(c);
137 
138 }
.fi

.PP
References \fBMAX_LOG_LEVEL\fP, and \fBMaxOS::common::OutputStream::write_char()\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Max OS from the source code\&.


kernel/src/common/logger.cpp
/**
 * @file logger.cpp
 * @brief Implements a Logger class for logging messages to multiple output streams
 *
 * @date 4th April 2025
 * @author Max Tyson
 */

#include <common/logger.h>
#include <stdarg.h>
#include <drivers/console/vesaboot.h>
#include <common/version.h>
#include <system/cpu.h>
#include <processes/scheduler.h>

using namespace MaxOS;
using namespace MaxOS::common;
using namespace MaxOS::drivers::console;
using namespace MaxOS::processes;
using namespace MaxOS::system;

/**
 * @brief Constructs a Logger object and sets it as the active logger
 */
Logger::Logger()
: m_log_writers()
{

	s_active_logger = this;

	// The following line is generated automatically by the MaxOS build system.
	s_progress_total = 22;

}


Logger::~Logger() = default;

/**
 * @brief Adds an output stream to the logger
 *
 * @param log_writer The output stream to add
 */
void Logger::add_log_writer(OutputStream* log_writer) {

	// If the list is not empty
	if (m_log_writer_count >= MAX_LOG_WRITERS)
		return;

	// Add the output stream to the list
	m_log_writers[m_log_writer_count] = log_writer;
	m_log_writers_enabled[m_log_writer_count] = true;
	m_log_writer_count++;

	// Print the setup info
	*this << LogLevel::INFO << "Logger setup: " << m_log_writer_count << "\n";
	*this << LogLevel::HEADER << "MaxOS v" << VERSION_STRING << " [ build " << BUILD_NUMBER << " ]\n";

}

/**
 * @brief Removes an output stream from the logger
 *
 * @param log_writer The output stream to remove
 */
void Logger::disable_log_writer(OutputStream* log_writer) {

	// If the list is empty
	if (m_log_writer_count == 0)
		return;

	// Find the output stream in the list
	for (int i = 0; i < m_log_writer_count; i++)
		if (m_log_writers[i] == log_writer)
			m_log_writers_enabled[i] = false;
}

/**
 * @brief Sets the log level of the logger
 *
 * @param log_level The log level to set
 */
void Logger::set_log_level(LogLevel log_level) {

	// Set the log level
	m_log_level = log_level;

	// Update the progress bar
	if (log_level == LogLevel::INFO) {
		VESABootConsole::update_progress_bar((m_progress_current * 100) / s_progress_total);
		m_progress_current++;
	}

	// Print the header
	switch (log_level) {

		case LogLevel::HEADER:
			*this << ANSI_COLOURS[ANSIColour::FG_Blue] << "[  BOOT    ] ";
			break;

		case LogLevel::INFO:
			*this << ANSI_COLOURS[ANSIColour::FG_Cyan] << "[  INFO    ]" << ANSI_COLOURS[ANSIColour::FG_White] << " ";
			break;

		case LogLevel::DEBUG:
			*this << ANSI_COLOURS[ANSIColour::FG_Yellow] << "[  DEBUG   ]" << ANSI_COLOURS[ANSIColour::Reset] << " ";
			break;

		case LogLevel::WARNING:
			*this << ANSI_COLOURS[ANSIColour::BG_Yellow] << ANSI_COLOURS[FG_White] << "[ WARNING  ]" << ANSI_COLOURS[ANSIColour::Reset] << " ";
			break;

		case LogLevel::ERROR:
			*this << ANSI_COLOURS[ANSIColour::BG_Red] << "[  ERROR   ]" << ANSI_COLOURS[ANSIColour::Reset] << " ";
			break;
	}

	GlobalScheduler::print_running_header();
}


/**
 * @brief Writes a character to the logger
 *
 * @param c The character to write
 */
void Logger::write_char(char c) {

	// Ensure logging at this level is enabled
	if (m_log_level > MAX_LOG_LEVEL)
		return;

	// Write the character to all output streams
	for (int i = 0; i < m_log_writer_count; i++)
		if (m_log_writers_enabled[i])
			m_log_writers[i]->write_char(c);

}

/**
 * @brief Gets the active logger
 *
 * @return The active logger
 */
Logger &Logger::Out() {

	return *active_logger();
}

/**
 * @brief Gets active logger set to task level
 *
 * @return The task logger
 */
Logger Logger::HEADER() {

	s_active_logger->set_log_level(LogLevel::HEADER);
	return Out();

}

/**
 * @brief Gets active logger set to info level
 *
 * @return The info logger
 */
Logger Logger::INFO() {

	s_active_logger->set_log_level(LogLevel::INFO);
	return Out();
}

/**
 * @brief Gets active logger set to DEBUG level
 *
 * @return The debug logger
 */
Logger Logger::DEBUG() {

	s_active_logger->set_log_level(LogLevel::DEBUG);
	return Out();

}

/**
 * @brief Gets active logger set to WARNING level
 *
 * @return The warning logger
 */
Logger Logger::WARNING() {

	s_active_logger->set_log_level(LogLevel::WARNING);
	return Out();
}

/**
 * @brief Gets active logger set to ERROR level
 *
 * @return The error logger
 */
Logger Logger::ERROR() {

	s_active_logger->set_log_level(LogLevel::ERROR);
	return Out();

}

/**
 * @brief Gets the active logger (to be used to modify the logger)
 *
 * @return The active logger
 */
Logger *Logger::active_logger() {
	return s_active_logger;
}

/**
 * @brief Prints a formatted string to the logger
 *
 * @param format The format string
 * @param ... The arguments to format
 */
void Logger::printf(char const *format, ...) {

	// Create a pointer to the data
	va_list parameters;
	va_start(parameters, format);

	write(String::formatted(format, parameters));
}

/**
 * @brief Puts the system into a panic state if the condition is false, printing the message first
 *
 * @param condition The condition to check if is met
 * @param message The message to print if the condition fails
 */
void Logger::ASSERT(bool condition, char const *message, ...) {

	// If the condition is met then everything is ok
	if (condition)
		return;

	va_list parameters;
	va_start(parameters, message);

	// Hang the system
	system::CPU::PANIC(String::formatted(message, parameters).c_str());
}

/**
 * @brief Sets the log level of the logger
 *
 * @param log_level The log level to set
 * @return This logger
 */
Logger &Logger::operator<<(LogLevel log_level) {

	set_log_level(log_level);
	return *this;

}